frame/treasury/src/lib.rs

```diff
@@ -133,10 +137,45 @@ pub struct Proposal<AccountId, Balance> {
 	bond: Balance,
 }
 
+/// The state of the payment claim.
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, MaxEncodedLen, RuntimeDebug, TypeInfo)]
+pub enum PaymentState<Id> {
+	/// Pending claim.
+	Pending,
+	/// Payment attempted with a payment identifier.
+	Attempted { id: Id },
+	/// Payment failed.
+	Failed,
+}
+
+/// Info regarding an approved treasury spend.
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, MaxEncodedLen, RuntimeDebug, TypeInfo)]
+pub struct SpendStatus<AssetKind, AssetBalance, Beneficiary, BlockNumber, PaymentId> {
+	// The kind of asset to be spent.
+	asset_kind: AssetKind,
+	/// The asset amount of the spend.
+	amount: AssetBalance,
+	/// The beneficiary of the spend.
+	beneficiary: Beneficiary,
+	/// The block number by which the spend has to be claimed.
+	expire_at: BlockNumber,
+	/// The status of the payout/claim.
+	status: PaymentState<PaymentId>,
+}
```

Comment:
In order to prevent a future migration, perhaps add a `valid_from: BlockNumber` field? The default would be `now` when `spend` executes, but I'd like to eventually support a recurring payments model, for example: Pay `amount` of `asset_kind` every 3 months for 4 times.

Then we can add a `void_claims` call with privileged origin. The user story here is that someone could make a treasury proposal with many milestones, but they don't want to spend their time lobbying governance for each step (but governance doesn't want to pay out up front). This would approve a set of payments that would be claimable in the future by default, but if the beneficiary isn't meeting the original expectations of the stakeholders, the future payments could be voided.

Reply:
sounds good to me.
a user can batch the `spend` call with different `valid_from`. the dispatchable already takes into consideration all the batched spends when asserting against the origin's maximum amount.

however, it's worth noting that if the `SpendPeriod` value has changed, it will affect not only the future spends but also all previously created spends waiting to be claimed.

Reply:
Yeah good point with just batching spends like that. I think the functionality in terms of new dispatchables / `SpendPeriod` management can come in later PRs as additions, but mostly want to ensure the datagrams are more futureproof.

Reply:
I am including the `valid_from` as an additional parameter to the new `spend` call and adding new to `void_spend` in the current PR

Reply:
updated

Reply:
I also store `expire_at`, the modification to `Config::SpendPeriod` does not affect approved spends